[core] Remove custom ov::optional

oToToT opened the pull request:
### Description:
As CPP17 is the default standard of OpenVINO, there is no need to use the self-implemented optional struct.

Note: discussed in https://github.com/openvinotoolkit/openvino/pull/28942.

### Related PRs:
- #28942

praasz commented:
build_jenkins

mlukasze commented:
build_jenkins

oToToT commented:
Fixed the build on Ubuntu24.04.

Found that the bool operator of ov::optional is not marked as explicit while std::optional is.

mlukasze commented:
build_jenkins

oToToT commented:
can't see jenkins failure. do i need to do anything?

praasz commented:
> can't see jenkins failures logs. do i need to do anything?

There are just build error which should be visible during standard local build of project.
```
 In file included from openvino/src/core/src/op/segment_max.cpp:10:

[2025-02-14T05:54:40.407Z] openvino/src/core/shape_inference/include/segment_max_shape_inference.hpp: In function ‘std::vector<UserT> ov::op::v16::shape_infer(const ov::op::v16::SegmentMax*, const std::vector<TShape>&, const ov::ITensorAccessor&)’:

[2025-02-14T05:54:40.408Z] openvino/src/core/shape_inference/include/segment_max_shape_inference.hpp:54:60: error: ‘optional’ is not a member of ‘ov’; did you mean ‘std::optional’?
```

oToToT commented:
Thanks for sharing the error message. It seems like these are some file not in my forked branch but in the latest trunk.

I've rebased and updated those files to reflect the removal of `ov::optional`.

praasz commented:
build_jenkins

mlukasze commented:
build_jenkins

mlukasze commented:
@itikhono please take a look, it's ready to merge, but would prefer to have your approval as well

dmitry-gorokhov reviewed (approved):
CPU part LGTM

oToToT commented:
rebased and fixed the merge conflict

mlukasze commented:
build_jenkins

mlukasze commented:
![image](https://github.com/user-attachments/assets/90ce53ad-3457-417b-abdd-f0b5bd337e41)

oToToT commented:
> ![image](https://private-user-images.githubusercontent.com/48962551/414569913-90ce53ad-3457-417b-abdd-f0b5bd337e41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5NDU3NDUsIm5iZiI6MTczOTk0NTQ0NSwicGF0aCI6Ii80ODk2MjU1MS80MTQ1Njk5MTMtOTBjZTUzYWQtMzQ1Ny00MTdiLWFiZGQtZjBiNWJkMzM3ZTQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE5VDA2MTA0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyMDYwMGZmMzdkNmM3MDIzZjNhOTJkYjRiNjJiMjZkMWY5MDcyNTIyNzdiYzZiMzU3NmRlODY4Y2Y4ZTFmNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.s_C9w2BrGqO2XBDsYGcIdbetZ1CcFi2jFPpk5k_CVkI)

oops. apologize for that. It should be fixed now. 

I've just built the code myself again to check it now.

mlukasze commented:
build_jenkins

mlukasze commented:
CI reported problem with clang-format, but that one is on us, do not change anything for now, please.
We will rebase and restart CI as soon as problem will be solved by our teams

mlukasze commented:
no problem, as rebase is done I will run CI as soon as we will have confirmation that problem has been solved.
stay tuned

mlukasze commented:
build_jenkins
